### PR TITLE
fix wmstats report for data update time

### LINKED ISSUE #8347: WMStats should report AgentStatusWatcher instead of AnalyticsDataCollector
I found that AgentStatusWatcher was down (while AnalyticsDataCollector was up), even though wmstats reports an error with `status: Data is not updated: AnalyticsDataCollector Down; `

which basically masks any other component that might be down but not getting reported by AgentStatusWatcher.

### PR DESCRIPTION
**Pull request by ticoann**
fixes #8347

### CODE REVIEW

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py [hunk 1]
```diff
@@ -142,10 +144,18 @@ def algorithm(self, parameters):
 
             self.localSummaryCouchDB.uploadData(requestDocs)
             logging.info("Request data upload success\n %s request, \nsleep for next cycle" % len(requestDocs))
-            DataUploadTime.setInfo(uploadTime, "ok")
+
+            self.centralWMStatsCouchDB.updateAgentInfoInPlace(self.agentInfo["agent_url"],
+                                                              {"data_last_update": uploadTime, "data_error": "ok"})
 
         except Exception as ex:
             logging.error("Error occurred, will retry later:")
```
Comment (amaltaro) [inline src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: How about replacing these error logging by one exception? traceback import can probably be removed then.

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py [hunk 2]
```diff
@@ -142,10 +144,18 @@ def algorithm(self, parameters):
...
             logging.error(str(ex))
-            DataUploadTime.setInfo(False, str(ex))
             logging.error("Trace back: \n%s" % traceback.format_exc())
+
+            try:
+                self.centralWMStatsCouchDB.updateAgentInfoInPlace(self.agentInfo["agent_url"], {"data_error": str(ex)})
```
Comment (amaltaro) [inline src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: Just trying to understand. If there was an exception above, we immediately post this error to couch wmstats, which is visible through the agent_info view? Does it show up like a component error or something else?

**Review by amaltaro — APPROVED** [review]
Patch looks good, Seangchan.
I'm just wondering whether these new key/values updated in place (data_*) will get overwritten/removed by AgentStatusWatcher updates?

I need to eventually look at this component and check what it does :-) It seems it collects some interesting data, which is never used and never looked by anyone though...

**Comment by ticoann** [comment]
> I'm just wondering whether these new key/values updated in place (data_*) will get overwritten/removed by AgentStatusWatcher updates?

That is good point, it will be removed every time AgentStatusWatcher updates. I will fix that.

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py (hunk 1 above)
Comment (ticoann) [reply src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: Thanks, I will change that.

File: src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py (hunk 2 above)
Comment (ticoann) [reply src/python/WMComponent/AnalyticsDataCollector/AnalyticsPoller.py]: Yes it will be visible in agent_info view. It should show up "data_error".

**Comment by ticoann** [comment]
@amaltaro, Alan I updated patch, Thanks.

**Comment by amaltaro** [comment]
Seangchan, sorry for the late review. If it's not much work, it might be easier and cleaner to do something as Erik did for the DrainStatusPoller thread, using a class method to access the AnalyticsDataCollector info from AgentStatusWatcher and making sure that that never gets overwritten.

See this code for instance:
https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/AgentStatusWatcher/DrainStatusPoller.py#L51

if you like it, then I think it's worth it changing this PR again :)

**Comment by ticoann** [comment]
Hi Alan, Actually that was the whole point of this patch (It was using class cache). 😄 But it doesn't work since those 2 components doesn't run in the same process, hence it doesn't share memory.

I could use the zmq for communicating between processes but I thought current patch is simpler.

**Pull request merged**